routers/repo/http.go

```diff
@@ -68,189 +52,188 @@ func HTTP(ctx *context.Context) {
-		authPasswd   string
-		environ      []string
-	)
+	var environ []string
 
-	// check access
-	if askAuth {
+	// Check access
+	// Only public pull don't need auth.
+	if repo.IsPrivate || !isPull || setting.Service.RequireSignInView {
+		var reverseProxyAuth bool
+		var authHeader = ctx.Req.Header.Get("Authorization")
 		if setting.Service.EnableReverseProxyAuth {
-			authUsername = ctx.Req.Header.Get(setting.ReverseProxyAuthUser)
-			if len(authUsername) == 0 {
-				ctx.HandleText(401, "reverse proxy login error. authUsername empty")
-				return
-			}
-			authUser, err = models.GetUserByName(authUsername)
-			if err != nil {
-				ctx.HandleText(401, "reverse proxy login error, got error while running GetUserByName")
-				return
-			}
-		} else {
-			authHead := ctx.Req.Header.Get("Authorization")
-			if len(authHead) == 0 {
-				ctx.Resp.Header().Set("WWW-Authenticate", "Basic realm=\".\"")
-				ctx.Error(http.StatusUnauthorized)
+			reverseProxyAuth = true
+			authHeader = ctx.Req.Header.Get(setting.ReverseProxyAuthUser)
+		}
+		authUser, promptForAuth, httpStatus, err := checkAccess(repo, unitType, isPull, authHeader, reverseProxyAuth)
+		if err != nil {
+			if httpStatus == http.StatusInternalServerError {
+				ctx.Handle(httpStatus, "checkAccess", err)
 				return
 			}
+			ctx.HandleText(httpStatus, err.Error())
+			return
+		}
+		if promptForAuth {
+			ctx.Resp.Header().Set("WWW-Authenticate", "Basic realm=\".\"")
+			ctx.Error(http.StatusUnauthorized)
+			return
+		}
 
-			auths := strings.Fields(authHead)
-			// currently check basic auth
-			// TODO: support digit auth
-			// FIXME: middlewares/context.go did basic auth check already,
-			// maybe could use that one.
-			if len(auths) != 2 || auths[0] != "Basic" {
-				ctx.HandleText(http.StatusUnauthorized, "no basic auth and digit auth")
-				return
-			}
-			authUsername, authPasswd, err = base.BasicAuthDecode(auths[1])
-			if err != nil {
-				ctx.HandleText(http.StatusUnauthorized, "no basic auth and digit auth")
-				return
-			}
+		environ = []string{
+			fmt.Sprintf("%s=%s", models.EnvRepoUsername, username),
+			fmt.Sprintf("%s=%s", models.EnvRepoName, reponame),
+			fmt.Sprintf("%s=%s", models.EnvPusherName, authUser.Name),
+			fmt.Sprintf("%s=%d", models.EnvPusherID, authUser.ID),
+			fmt.Sprintf("%s=%d", models.ProtectedBranchRepoID, repo.ID),
+			fmt.Sprintf("%s=%t", models.EnvRepoIsWiki, unitType == models.UnitTypeWiki),
+		}
+	}
 
-			authUser, err = models.UserSignIn(authUsername, authPasswd)
-			if err != nil {
-				if !models.IsErrUserNotExist(err) {
-					ctx.Handle(http.StatusInternalServerError, "UserSignIn error: %v", err)
-					return
-				}
-			}
+	HTTPBackend(ctx, &serviceConfig{
+		UploadPack:  true,
+		ReceivePack: true,
+		Env:         environ,
+	})(ctx.Resp, ctx.Req.Request)
+}
+
+func checkAccess(repo *models.Repository, unitType models.UnitType, isPull bool, authHeader string, reverseProxyAuth bool) (*models.User, bool, int, error) {
```

Comment:
maybe become a middleware

Reply:
Checking access is usually somewhat specific to what you're checking. E.g. here we need to pass `isPull` which would be useless in 99.9% of all other handlers. Trying to "shoehorn" things into middlewares for the sake of reusability is what leads to having to refactor code. (Like this PR does... since IMO it should be different handlers for `upload-pack` and `receive-pack`... but nvm that)

Reply:
OTOH, if we break this function down some parts could easily be middlewares... like ripping out the `authHeader, reverseProxyAuth` into it's on part (which I think we have somewhere...)